extractor.go

```diff
@@ -186,87 +188,112 @@ func (e *Extractor) Tar(ctx context.Context, body io.Reader, location string, re
 
 // Zip extracts a .zip archived stream of data in the specified location.
 // It accepts a rename function to handle the names of the files (see the example).
-func (e *Extractor) Zip(ctx context.Context, body io.Reader, location string, rename Renamer) error {
+func (e *Extractor) Zip(originalCtx context.Context, body io.Reader, location string, rename Renamer) error {
 	// read the whole body into a buffer. Not sure this is the best way to do it
 	buffer := bytes.NewBuffer([]byte{})
-	copyCancel(ctx, buffer, body)
+	copyCancel(originalCtx, buffer, body)
 
 	archive, err := zip.NewReader(bytes.NewReader(buffer.Bytes()), int64(buffer.Len()))
 	if err != nil {
 		return errors.Annotatef(err, "Read the zip file")
 	}
 
-	links := []link{}
-
-	// We make the first pass creating the directory structure, or we could end up
-	// attempting to create a file where there's no folder
-	for _, header := range archive.File {
-		select {
-		case <-ctx.Done():
-			return errors.New("interrupted")
-		default:
+	fileChan := make(chan *zip.File)
+	go func() {
```

Comment:
Should we run this function under the same `errgroup`? I'd create the `errgroup` before this one and run it with `g.Go(...)`

Reply:
Yes, didn't think about it.